File: frontend/src/services/api/relativeValuation.ts
import api from './client'

export interface SectorInfo {
  key: string
  label: string
}

export interface UniverseStock {
  code: string
  name: string
  sector: string
  sector_label: string
}

export interface StockMetric {
  code: string
  name: string
  market: string
  price: number
  change_pct?: number | null
  market_cap?: number | null
  pe?: number | null
  pb?: number | null
  ps?: number | null
  dividend_yield?: number | null
  roe?: number | null
  net_margin?: number | null
  revenue_growth?: number | null
  profit_growth?: number | null
  pe_pct?: number | null
  pb_pct?: number | null
  ps_pct?: number | null
  div_pct?: number | null
  pe_dev?: number | null
  pb_dev?: number | null
  ps_dev?: number | null
  div_dev?: number | null
  attractiveness?: number | null
  rating?: string
  rating_level?: string
}

export interface CompareResult {
  market: string
  sector: string
  sector_label: string
  stocks: StockMetric[]
  medians: Record<string, number | null>
  count: number
  data_note: string
  error?: string
  target?: StockMetric | null
  target_code?: string
  note?: string
}

export const relativeValuationApi = {
  /** 获取 A/H 市场可对比的行业清单 */
  getSectors: () => api.get<{ A: SectorInfo[]; HK: SectorInfo[] }>('/relative-valuation/sectors').then(r => r.data),

  /** 同一行业组内对比 */
  compareSector: (market: string, sector: string) =>
    api
      .get<CompareResult>('/relative-valuation/compare', { params: { market, sector } })
      .then(r => r.data),

  /** 单只股票 vs 其所在行业同业 */
  compareStock: (market: string, code: string, sector?: string) =>
    api
      .get<CompareResult>('/relative-valuation/stock', { params: { market, code, sector } })
      .then(r => r.data),

  /** 可搜索标的清单（代码/名称/行业），供「选一只标的」自动完成 */
  getStocks: (market: string) =>
    api
      .get<{ market: string; stocks: UniverseStock[] }>('/relative-valuation/stocks', { params: { market } })
      .then(r => r.data),
}

File: backend/app/api/relative_valuation.py
"""相对估值法 API — 同行业 A股/港股 跨市场相对估值对比

端点：
  GET /api/relative-valuation/sectors   返回 A/H 市场可选行业清单
  GET /api/relative-valuation/stocks?market=A   返回可搜索标的清单（代码/名称/行业）
  GET /api/relative-valuation/compare?market=A&sector=银行   组内同业对比
  GET /api/relative-valuation/stock?market=A&code=600519&sector=白酒   单只 vs 同业
"""

from fastapi import APIRouter, Query

from app.services import relative_valuation_service as rv

router = APIRouter()


@router.get("/sectors")
def get_sectors():
    """返回 A/H 市场可对比的行业清单。"""
    return rv.get_sectors()


@router.get("/stocks")
def get_stocks(market: str = Query(..., description="市场: A 或 HK")):
    """返回该市场可搜索的标的清单（代码/名称/行业），供前端「选一只标的」自动完成。"""
    return rv.get_stock_universe(market)


@router.get("/compare")
def compare_sector(
    market: str = Query(..., description="市场: A 或 HK"),
    sector: str = Query(..., description="行业 key（来自 /sectors）"),
):
    """同一行业内各股票的 PE/PB/PS/股息率 组内分位、相对中位数偏离与综合吸引力评级。"""
    return rv.compare_sector(market, sector)


@router.get("/stock")
def compare_stock(
    market: str = Query(..., description="市场: A 或 HK"),
    code: str = Query(..., description="股票代码"),
    sector: str = Query(None, description="行业 key；不填则自动定位"),
):
    """单只股票与其所在行业已知同业的相对估值对比。"""
    return rv.compare_stock(market, code, sector)
